## Mean, Median, and Mode

In Machine Learning (and in mathematics) there are often three values that interests us:

### .Mean - The average value
### .Median - The mid point value
### .Mode - The most common value

In [1]:
# Example: We have registered the speed of 13 cars:

speed = [99,86,87,88,111,86,103,87,94,78,77,85,86]

mean_speed = sum(speed)/len(speed)
print(mean_speed)

89.76923076923077


## The NumPy module has a method for this.

In [3]:
# Use the NumPy mean() method to find the average speed:
import numpy as np
mean_speed = np.mean(speed)
print(f"mean_speed: {mean_speed}")

mean_speed: 89.76923076923077


In [5]:
# Use the NumPy median() method to find the middle value:
import numpy as np
median_speed = np.median(speed)
print(f"median_speed: {median_speed}")

median_speed: 87.0


In [7]:
# Use the SciPy mode() method to find the number that appears the most:
from scipy import stats
mode_speed = stats.mode(speed)
print(f"mode_speed: {mode_speed}")

mode_speed: ModeResult(mode=np.int64(86), count=np.int64(3))


# Standard Deviation?
Standard deviation is a number that describes how spread out the values are.

A low standard deviation means that most of the numbers are close to the mean (average) value.

A high standard deviation means that the values are spread out over a wider range.

In [9]:
# Example: We have registered the speed of 13 cars:
speed = [99,86,87,88,111,86,103,87,94,78,77,85,86]

mean_speed = np.mean(speed)
print(f"mean_speed: {mean_speed}")

# Calculate the standard deviation of the speed:
std_speed = np.std(speed)
print(f"std_speed: {std_speed}")

mean_speed: 89.76923076923077
std_speed: 9.258292301032677


Meaning that most of the values are within the range of 9.26 from the mean value, which is 80.50.

As you can see, a higher standard deviation indicates that the values are spread out over a wider range.

The NumPy module has a method to calculate the standard deviation:

In [10]:
speed = [32,111,138,28,59,77,97]


mean_speed = np.mean(speed)
print(f"mean_speed: {mean_speed}")

# Calculate the standard deviation of the speed:
std_speed = np.std(speed)
print(f"std_speed: {std_speed}")

mean_speed: 77.42857142857143
std_speed: 37.84501153334721


Meaning that most of the values are within the range of 37.85 from the mean value, which is 77.4.

As you can see, a higher standard deviation indicates that the values are spread out over a wider range.

The NumPy module has a method to calculate the standard deviation:

##  Variance is another number that indicates how spread out the values are.

### In fact, if you take the square root of the variance, you get the standard deviation!

### the other way around, if you multiply the standard deviation by itself, you get the variance!

### To calculate the variance you have to do as follows:

In [11]:
import numpy as np
from numpy import mean, median, std
speed = [32,111,138,28,59,77,97]

In [12]:
# First Find mean value
mean_value = np.mean(speed)
print(f"mean_value: {mean_value}")

mean_value: 77.42857142857143


In [14]:
# Calculate the variance of the speed:
variance = np.var(speed)
print(f"variance: {variance}")

variance: 1432.2448979591834


## Steps to Calculate Variance

In [17]:
# first find mean value
mean_value = np.sum(speed)/len(speed)
# For each value find difference from mean value, then square it
squared_diff = [(x - mean_value)**2 for x in speed]
for i in range(len(speed)):
    print(f"speed: {speed[i]}, mean_value: {mean_value}, squared_diff: {squared_diff[i]}")
    
# Then find the average of those squared differences    
squared_diff_sum = sum(squared_diff)
variance = squared_diff_sum/len(speed)
print(f"variance: {variance}")

speed: 32, mean_value: 77.42857142857143, squared_diff: 2063.7551020408164
speed: 111, mean_value: 77.42857142857143, squared_diff: 1127.0408163265304
speed: 138, mean_value: 77.42857142857143, squared_diff: 3668.8979591836733
speed: 28, mean_value: 77.42857142857143, squared_diff: 2443.183673469388
speed: 59, mean_value: 77.42857142857143, squared_diff: 339.6122448979593
speed: 77, mean_value: 77.42857142857143, squared_diff: 0.18367346938775683
speed: 97, mean_value: 77.42857142857143, squared_diff: 383.04081632653055
variance: 1432.2448979591834


# Calculate Standard with numpy std function

In [18]:
import numpy as np

speed = [32,111,138,28,59,77,97]

x = np.std(speed)

print(x)

37.84501153334721


##  Symbols
Standard Deviation is often represented by the symbol Sigma: σ

Variance is often represented by the symbol Sigma Squared: σ2

##  Percentiles
Percentiles are used in statistics to give you a number that describes the value that a given percent of the values are lower than.

Example: Let's say we have an array that contains the ages of every person living on a street.

In [20]:
# ages of 21 persons who attended a party:
# What is the 75. percentile ages of the persons?
ages = [5,31,43,48,50,41,7,11,15,39,80,82,32,2,8,6,25,36,27,61,31]
percentile_75 = np.percentile(ages, 75)
print(f"75th percentile age: {percentile_75}, which means that 75% of the ages are below this value.")

75th percentile age: 43.0, which means that 75% of the ages are below this value.


In [21]:
# Use the NumPy percentile() method to find the 75. percentile ages of the persons:
ages = [5,31,43,48,50,41,7,11,15,39,80,82,32,2,8,6,25,36,27,61,31]
# What is the 75. percentile ages of the persons?
percentile_75 = np.percentile(ages, 75)
print(f"75th percentile age: {percentile_75}, which means that 75% of the ages are below this value.")

75th percentile age: 43.0, which means that 75% of the ages are below this value.


In [22]:
# Use the NumPy percentile() method to find the 75. percentile ages of the persons:
ages = [5,31,43,48,50,41,7,11,15,39,80,82,32,2,8,6,25,36,27,61,31]
# What is the 95. percentile ages of the persons?
percentile_95 = np.percentile(ages, 95)
print(f"95th percentile age: {percentile_95}, which means that 95% of the ages are below this value.")

95th percentile age: 80.0, which means that 95% of the ages are below this value.
